### PR TITLE
ONME-3981 DNS query should prefer network interface IP version

### PR DESCRIPTION
**Pull request by tymoteuszblochmobica**
### Description
ONME-3981 DNS query should prefer network interface IP version
This was raised as a defect from Cellular team.
The test environment was LWIP with DEFAULT stack supporting both IP4 with DHCP and ND6 IP6
In result LWIP  network interface has both IP version addresses.  Test environment network infrastructure supports both IP4 and IP6.
 IP preferred version   was hardcoded to IP4 and based on MBED CONFIG.
Therefore DNS query returns always IP4.
Proposed extension adds  default parameter eg. IP_type(4/6)  to connect() eg.  connect(IP_type = DEFAULT_STACK) member of  NetworkInterface and derived classes.  Now connect(IP_type ) member called from user app can force IP_Type of interface  or  can be left unchanged so original implementation supporting both IP versions will be used. However dual stack is not supported by Mbed.

This is simple experimental, proposal.  Final solution should be a subject of separate discussion and design 

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### CODE REVIEW

**Comment by tymoteuszblochmobica** [comment]
@seppotakalo please review

**Comment by tymoteuszblochmobica** [comment]
@mikaleppanen please review

**Comment by tymoteuszblochmobica** [comment]
@kjbracey-arm please review

File: features/netsocket/NetworkInterface.h [hunk 1]
```diff
@@ -147,7 +147,7 @@ class NetworkInterface: public DNS {
      *  @return  NSAPI_ERROR_IS_CONNECTED if interface is already connected.
      *  @return  negative error code on failure.
      */
-    virtual nsapi_error_t connect() = 0;
```
Comment (SeppoTakalo) [inline features/netsocket/NetworkInterface.h:140]: I don't feel that this is the correct API change.

`NetworkInterface::connect()` has traditionally taken only configuration parameters and the non-parameter version is equivalent between all interfaces.

Now we are changing the behaviour of the non-parameter version, but not those that take parameters.


Can you point out the place in DNS resolver that utilises this knowledge of preferred IP stack version?

**Comment by SeppoTakalo** [comment]
So,

Within the [nsapi_dns_query_multiple()](https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/nsapi_dns.cpp#L435) we already get pointer to `NetworkStack`

Why cannot we query its IP address by calling [const char *get_ip_address()](https://github.com/ARMmbed/mbed-os/blob/master/features/netsocket/NetworkStack.h#L46)?

Then try [stoip4()](https://github.com/ARMmbed/mbed-os/blob/master/features/frameworks/nanostack-libservice/mbed-client-libservice/ip4string.h#L43) and [stoip6()](https://github.com/ARMmbed/mbed-os/blob/master/features/frameworks/nanostack-libservice/mbed-client-libservice/ip6string.h#L64) to detect whether the returned IP address was IPv4 or IPv6.

Then we could prefer those IP addresses that are in same family as what the stack returns first.

**Pull request closed without merging**